theseus/labs/lie/lie_tensor.py

```diff
@@ -0,0 +1,272 @@
+        return self._unary_jop_base(tangent_vector, self._fn_lib.exp)
+
+    def jlog(self) -> _JFnReturnType:
+        return self._unary_jop_base(self._t, self._fn_lib.log, out_is_group=False)
+
+    def jinv(self) -> _JFnReturnType:
+        return self._unary_jop_base(self._t, self._fn_lib.inv)
+
+    def jcompose(self, other: "LieTensor") -> _JFnReturnType:
+        self._check_ltype(other, "jcompose")
+        jacs: List[torch.Tensor] = []
+        op_res = self.new(self._fn_lib.compose(self._t, other._t, jacobians=jacs))
+        return jacs, op_res
+
+    def _no_jop(self, input0: TensorType) -> _JFnReturnType:
+        raise NotImplementedError
+
+    jadjoint = _no_jop
+    jhat = _no_jop
+    jvee = _no_jop
+    jlift = _no_jop
+    jproject = _no_jop
+    jleft_act = _no_jop
+    jleft_project = _no_jop
+
+
+# ----------------------------
+# Tensor creation functions
+# ----------------------------
+def new(
+    data: Any,
+    ltype: Optional[_ltype] = None,
+    dtype: torch.dtype = torch.float,
+    device: Union[torch.device, str] = None,
+) -> "LieTensor":
+    if isinstance(data, LieTensor):
+        return data.new(data)
+    if ltype is None:
+        raise ValueError("ltype must be provided.")
+    if not isinstance(data, torch.Tensor):
+        data = torch.as_tensor(data, dtype=dtype, device=device)
+    return LieTensor(data, ltype=data.ltype)
+
+
+as_lietensor = new
+
+
+# Similar to the one in functional, except it returns a LieTensor
+# and receives a ltype
+class _RandFnType(Protocol):
+    def __call__(
+        self,
+        ltype: _ltype,
+        *size: int,
+        generator: Optional[torch.Generator] = None,
+        dtype: Optional[torch.dtype] = None,
+        device: DeviceType = None,
+        requires_grad: bool = False,
+    ) -> LieTensor:
+        pass
+
+
+def _build_random_fn(op_name: str) -> _RandFnType:
+    assert op_name in ["rand", "randn"]
+
+    def fn(
+        ltype: _ltype,
+        *size: int,
+        generator: Optional[torch.Generator] = None,
+        dtype: Optional[torch.dtype] = None,
+        device: DeviceType = None,
+        requires_grad: bool = False,
+    ) -> LieTensor:
+        fn = {SE3: getattr(_se3_base, op_name), SO3: getattr(_so3_base, op_name)}[ltype]
+        return LieTensor(
+            fn(
+                *size,
+                generator=generator,
+                dtype=dtype,
+                device=device,
+                requires_grad=requires_grad,
+            ),
+            ltype=ltype,
+        )
+
+    return fn
+
+
+rand: _RandFnType = _build_random_fn("rand")
+randn: _RandFnType = _build_random_fn("randn")
+
+
+def log(group: LieTensor) -> torch.Tensor:
+    return group.log()
+
+
+def adj(group: LieTensor) -> torch.Tensor:
+    return group.adj()
+
+
+def inv(group: LieTensor) -> LieTensor:
+    return group.inv()
+
+
+def exp(ltype: _ltype, tangent_vector: torch.Tensor) -> "LieTensor":
```

Comment:
Curious if there is any standard convention on param first vs data first in a function call.

Reply:
In torch, the dtype is usually an optional keyword for any torch creation ops; however, for us the `ltype` is not optional, so it's a bit different. I don't have a strong preference if should go first or last.

Reply:
Similar to the other comment, we can follow norms from torch for what a user would expect regarding which goes first.

Reply:
> In torch, the dtype is usually an optional keyword for any torch creation ops; however, for us the `ltype` is not optional, so it's a bit different. I don't have a strong preference if should go first or last.

For generality, we can treat pytorch tensor as a Lie group as well. Maybe we can create another Lie group class called `Matrix` that are essentially torch tensors.

Reply:
Addressed the `ltype` arg order in 5a72342ccb6035db2054d038e9283650e3ac54c6.